### DISCUSSION TITLE
(bank): Errors In Third Party Tools Regarding LZSS (tested on 'legacy' and 'legacy/fastlzss')

### DISCUSSION THREAD

**Discussion by DevScarabyte** [FlipperPlz/BisUtils#19] [General]
**Describe the bug**
Legacy Branch since that's the branch I am using.

LZSS Compression error when packing, server doesn't read the PBO and Mikero can't open it.

I even applied the Fast LZSS from the rewrite and made it work without errors (check my fork) same error though.

**To Reproduce**
1. Pull Legacy from my fork
2. Build DLL and use with a packer
3. Run Server with new PBO or try to open

**Expected behavior**
PBO is read and running on server or opened by tools.

**Screenshots**
[If applicable, add screenshots to help explain your problem.](https://i.imgur.com/Y1EGNoF.png)

**Desktop (please complete the following information):**
 - OS: Windows

**Comment by FlipperPlz** [comment]
AFAIK compression is not the issue with the legacy version of BisUtils. The PBO implementation used in legacy is at a broken point of no return hence why I abandoned it, It doesnt write the correct checksum. To verify can you please try unpacking the PBO with a tool like PBOSpy or PBOManager and ensure all files have been correctly written?

**Comment by FlipperPlz** [comment]
Will work on adding write support to the new library today for everyone. Sorry for any inconveniences! The rewrite should be pretty stable as of now. Working on a few things for reading scrambled pbos right now.

**Comment by DevScarabyte** [comment]
If I disable compression the PBO works perfectly fine,

![Example](https://github.com/FlipperPlz/BisUtils/assets/1756242/c66804ed-eb8d-46c6-b172-bff73f1961c0)

The rewrite is usable :D! How exactly will it have changed from creating PBOs from a folder? 
This is how I currently do it using legacy
`var pbo = new PboFile(
    newPBO,
    PboFileOption.Create
);`

`
pbo.AddEntry(new PboDataEntryDto(pbo, new MemoryStream(System.Text.Encoding.UTF8.GetBytes(pd.data)), null, Compression) {
    EntryName = pd.name
}, true);
`

**Comment by FlipperPlz** [comment]
Currently, the provided implementation is primarily suitable for reading purposes. However, I am planning to work on enhancing the writing capabilities later today. At the moment, there are limited utilities available for conveniently reading or creating a` RVBank` object. Nevertheless, the signature of the `RVBank` constructor is as follows:
```csharp
RVBank::RVBank(
    string filename,
    BisBinaryReader reader,
    RVBankOptions options,
    Stream? synchronizeTo,
    ILogger? logger
)
```

- filename: This parameter represents the name of the file and is utilized to set a default prefix if no entry for the version is found.
- reader: It is an instance of the `BisBinaryReader` class, which is responsible for reading operations in the context of the `RVBank` object.
- options: This parameter is a list of options that can be utilized for both reading and writing operations on the PBO file.
- synchronizeTo: This parameter represents a stream to which data will be saved when the `BisSynchronizable<?>::SynchronizeWithStream()` method is called. It is optional and can be set to `null` if saving is not required.
- logger: It is an instance of any logger class that you wish to utilize. You can use `NullLogger.Instance` or set it to `null` to disregard logging.
An example options object for reading obfuscated banks looks like this (settings get added alot so this will change)
```csharp
var options = new RVBankOptions()
        {
            AllowObfuscated = true,
            FlatRead = false,
            RemoveBenignProperties = true,
            AllowMultipleVersion = true,
            IgnoreValidation = true,
            DecompressionTimeout = timeout!,
            Charset = Encoding.UTF8,
            RequireValidSignature = false,
            RegisterEmptyEntries = false,
            CompressionErrorsAreWarnings = false,
            RequireEmptyVersionMeta = false,
            RequireVersionMimeOnVersion = false,
            RequireFirstEntryIsVersion = false,
            RequireVersionNotNamed = false,
            AllowVersionMimeOnData = true,
            IgnoreInvalidStreamSize = true,
            AlwaysSeparateOnDummy = true,
            AllowUnnamedDataEntries = true,
            AllowDuplicateFileNames = true,
            RespectEntryOffsets = false,
            WriteValidOffsets = false,
            AsciiLengthTimeout = 510,
            AllowEncrypted = true
        };
```

**Comment by FlipperPlz** [comment]
> If I disable compression the PBO works perfectly fine,


I will be looking into this then shortly as I'm using the same compression in my rewrite. Thanks for the heads up
